docs/hugo/content/design/ADR-2024-07-Dynamic-Export-To-Secret-Or-ConfigMap.md

```diff
@@ -0,0 +1,428 @@
+
+**Secrets**
+```yaml
+  operatorSpec:
+    secrets:
+      hostName:
+        name: redis-secret
+        key: hostName
+      sslPort:
+        name: redis-secret
+        key: port
+    formattedSecrets:
+      values:
+        - name: redis-secret # Name of the destination secret
+          key: hostPort # Name of the key in the secret
+          value: "'%s:%s'.format([secret.hostName, secret.sslPort])" # Value (format) of the secret
+```
+
+**ConfigMaps**
+```yaml
+  operatorSpec:
+    formattedConfigMaps:
+      values:
+        - name: account-data # Name of the destination configMap
+          key: id # Name of the key in the configMap
+          value: "resource.status.id" # Value (format) of the configMap entry
+```
+
+**Pros**
+
+* Fits into our existing model relatively well.
+
+**Cons**
+
+
+#### Option 3: A new CRD `Exporter`
+
+The final name of this resource is TBD.
+
+```yaml
+apiVersion: serviceoperator.azure.com/v1
+kind: Exporter
+spec:
+  connectedResources:
+    - ref:
+        group: documentdb.azure.com
+        kind: DatabaseAccount
+        name: mydb
+      secrets:
+        - name: redis-secret # Name of the destination secret
+          key: hostPort # Name of the key in the secret
+          value: "'%s:%s'.format([secret.hostName, secret.sslPort])" # Value (format) of the secret
+      configMaps:
+        - name: account-data # Name of the destination configMap
+          key: id # Name of the key in the configMap
+          value: "resource.status.id" # Value (format) of the configMap entry
+```
+
+Open question if this would also support the well-typed approach we have now with fields like:
+```yaml
+    secrets:
+      hostName:
+        name: redis-secret
+        key: hostName
+      sslPort:
+        name: redis-secret
+        key: port
+```
+
+or if everything required a dynamic expression.
+
+**Pros**
+
+* Could combine multiple resources data into a single ConfigMap, though this brings questions of ownership
+  and deletion/propagation.
+
+**Cons**
+
+* Another way to do things. Possibly confusing because there will be 2 ways to export data to configmap/secret 
+  for many existing resources.
+* More complicated, especially setting up watches on N resource types so that changes are correctly monitored and
+  propagated quickly. That additional complexity doesn't necessarily bring all that many wins.
+* Access to secret data is more difficult to implement. We would need to have a static mapping of resource types and
+  the `Exporter` would need to call each referenced resources `ExportKubernetesResources`, and we'd likely need to 
+  refactor `ExportKubernetesResources` some to support the new pattern (though we may need to do this anyway).
+
+### Access to secret data
+
+Getting access to data to export to ConfigMaps is pretty easy, we'll just pass the whole `resource` to
+the expression. As shown elsewhere fields will be accessible in expressions like `resource.status.id`.
```

Comment:
There's also metadata though so I think just passing the whole top level object might be cleaner?

Reply:
I think the tension here is between _making things possible_ and _making common things easy_. 

Passing just the top level object makes things _possible_ but it forces every usage to incur a little bit of overhead. Passing multiple objects is more complex to implement, but has the benefit of making most usage simpler.

Reply:
Still not sure I agree. `resource.spec.whatever` isn't much more complicated than `spec.whatever`.


The advantages to us using top level names we choose are:
1. It disambiguates between the resource and its secrets, since for secret usage we'll support `secrets.myKey`. If we control the top-level name of all the inputs (the main 2 right now being `resource` and `secrets`), then we can ensure there are no collisions because we own the top of the path for each input. If we instead snip that part out (for just resource? or for secrets too?) then we run the risk (though it's small) that a top level field such as `secrets` collides.
2. It puts us in a better position to grow in the future by adding other top-level types going forward, such as `operator` (holding operator configuration?) or `cluster` (holding cluster details?). I don't know if we're ever going to want to add these but if we did, disambiguating between them at the top level helps with clarity.
3. It means we're consistent between `resource.x` and `secrets.y` in terms of the pattern, as opposed to doing `x` and `secrets.y`. I think including `secrets` as a prefix for secret values which are not on the resource is very useful from a security/safety point of view because we're making it really obvious the value you're about to output is a secret. Obviously we only support this in the context of saving to secrets so it's not like there's risk of exposing a secret value as plaintext but still the clarity seems a win for something sensitive like secrets.

I'll update the document to discuss this and leave it as an option question at the end of the document.